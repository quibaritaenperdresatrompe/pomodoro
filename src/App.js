import { makeStyles } from "@material-ui/styles";
import React, { useState } from "react";

import EmojiButton from "./EmojiButton";
import Horloge from "./Horloge";
import nextItem from "./nextItem";
import ThemeButton from "./ThemeButton";
import ThemeProvider from "./ThemeProvider";
import sessions from "./sessions";
import useTimer from "./useTimer";
import useAlarm from "./useAlarm";

const useStyles = makeStyles(() => ({
  root: {
    alignItems: "center",
    backgroundColor: ({ theme }) => (theme === "light" ? "white" : "black"),
    display: "flex",
    flexDirection: "column",
    justifyContent: "center",
    minHeight: "100vh",
    width: "100%",
  },
  header: {
    marginBottom: 16,
    minHeight: 36,
    width: "100%",
    color: "inherit",
  },
  toolbar: {
    alignItems: "center",
    display: "flex",
    justifyContent: "flex-end",
    padding: 16,
  },
  content: {
    alignItems: "center",
    display: "flex",
    flexDirection: "column",
    flexGrow: 1,
    justifyContent: "center",
    width: "100%",
  },
}));

const initType = "work";

export default function App() {
  const [theme, setTheme] = useState("light");
  const classes = useStyles({ theme });
  const [muted, setMuted] = useState(false);
  const [type, setType] = useState(initType);
  const [emoji, setEmoji] = useState(initType);
  const [setVolume, play, stop] = useAlarm();
  const timer = useTimer(sessions[type].duration, play);

  function onToggleSound() {
    const nextMuted = !muted;
    setMuted(nextMuted);
    setVolume(nextMuted ? 0 : 1);
  }

  function onChangeEmoji() {
    setEmoji(nextItem(emoji, sessions[type].emojis));
  }

  function onNewSession() {
    stop();
    const nextType = nextItem(type, Object.keys(sessions));
    setEmoji(nextType);
    setMuted(false);
    setType(nextType);
  }

  function onToggleTheme() {
    setTheme(theme === "light" ? "dark" : "light");
  }

  return (
    <ThemeProvider value={theme}>
      <div className={classes.root}>
        <header className={classes.header}>
          <div className={classes.toolbar}>
            <ThemeButton onClick={onToggleTheme} />
            <EmojiButton
              emoji={muted ? "ðŸ”•" : "ðŸ””"}
              label="Toggle sound"
              onClick={onToggleSound}
            />
          </div>
        </header>
        <main className={classes.content}>
          <Horloge
            time={timer}
            onClick={onNewSession}
            color={sessions[type].color}
          />
          <EmojiButton emoji={emoji} onClick={onChangeEmoji} size="large" />
        </main>
      </div>
    </ThemeProvider>
  );
}
