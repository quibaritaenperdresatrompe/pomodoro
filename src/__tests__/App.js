import { act, fireEvent, render } from "@testing-library/react";
import React from "react";

import App from "../App";
jest.mock("../useTimer", () => () => 60);
jest.mock("../useAlarm", () => () => [jest.fn(), jest.fn(), jest.fn()]);

describe(App.name, () => {
  beforeAll(() => {
    Object.defineProperty(window, "matchMedia", {
      value: jest.fn(() => {
        return {
          matches: true,
          addEventListener: jest.fn(),
          removeEventListener: jest.fn(),
        };
      }),
    });
  });

  it("renders correctly", () => {
    const { asFragment, getByText } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
    act(() => {
      fireEvent.click(getByText("ðŸ””"));
    });
    expect(asFragment()).toMatchSnapshot();
    act(() => {
      fireEvent.click(getByText("ðŸ”•"));
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly after emoji has been changed", () => {
    const { asFragment, getByText } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
    act(() => {
      fireEvent.click(getByText("work"));
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly after session type has been changed", () => {
    const { asFragment, getByText } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
    act(() => {
      fireEvent.click(getByText("01:00"));
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
