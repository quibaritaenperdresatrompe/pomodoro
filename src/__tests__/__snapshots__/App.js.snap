// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1 makeStyles-root-5"
  >
    <header
      class="makeStyles-header-2"
    >
      <div
        class="makeStyles-toolbar-3"
      >
        <button
          class="makeStyles-button-6"
          tooltiptext="Toggle theme"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle theme"
            class="makeStyles-emoji-7 makeStyles-emoji-8"
            role="img"
          >
            ☀️
          </span>
        </button>
        <button
          class="makeStyles-button-6"
          tooltiptext="Toggle sound"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle sound"
            class="makeStyles-emoji-7 makeStyles-emoji-9"
            role="img"
          >
            🔔
          </span>
        </button>
      </div>
    </header>
    <main
      class="makeStyles-content-4"
    >
      <code
        class="makeStyles-time-10 makeStyles-time-11"
      >
        01:00
      </code>
      <button
        class="makeStyles-button-6"
      >
        <span
          aria-hidden="true"
          aria-label=""
          class="makeStyles-emoji-7 makeStyles-emoji-12"
          role="img"
        >
          work
        </span>
      </button>
    </main>
  </div>
</DocumentFragment>
`;

exports[`App renders correctly 2`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1 makeStyles-root-5"
  >
    <header
      class="makeStyles-header-2"
    >
      <div
        class="makeStyles-toolbar-3"
      >
        <button
          class="makeStyles-button-6"
          tooltiptext="Toggle theme"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle theme"
            class="makeStyles-emoji-7 makeStyles-emoji-8"
            role="img"
          >
            ☀️
          </span>
        </button>
        <button
          class="makeStyles-button-6"
          tooltiptext="Toggle sound"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle sound"
            class="makeStyles-emoji-7 makeStyles-emoji-9"
            role="img"
          >
            🔕
          </span>
        </button>
      </div>
    </header>
    <main
      class="makeStyles-content-4"
    >
      <code
        class="makeStyles-time-10 makeStyles-time-11"
      >
        01:00
      </code>
      <button
        class="makeStyles-button-6"
      >
        <span
          aria-hidden="true"
          aria-label=""
          class="makeStyles-emoji-7 makeStyles-emoji-12"
          role="img"
        >
          work
        </span>
      </button>
    </main>
  </div>
</DocumentFragment>
`;

exports[`App renders correctly 3`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-1 makeStyles-root-5"
  >
    <header
      class="makeStyles-header-2"
    >
      <div
        class="makeStyles-toolbar-3"
      >
        <button
          class="makeStyles-button-6"
          tooltiptext="Toggle theme"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle theme"
            class="makeStyles-emoji-7 makeStyles-emoji-8"
            role="img"
          >
            ☀️
          </span>
        </button>
        <button
          class="makeStyles-button-6"
          tooltiptext="Toggle sound"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle sound"
            class="makeStyles-emoji-7 makeStyles-emoji-9"
            role="img"
          >
            🔔
          </span>
        </button>
      </div>
    </header>
    <main
      class="makeStyles-content-4"
    >
      <code
        class="makeStyles-time-10 makeStyles-time-11"
      >
        01:00
      </code>
      <button
        class="makeStyles-button-6"
      >
        <span
          aria-hidden="true"
          aria-label=""
          class="makeStyles-emoji-7 makeStyles-emoji-12"
          role="img"
        >
          work
        </span>
      </button>
    </main>
  </div>
</DocumentFragment>
`;

exports[`App renders correctly after emoji has been changed 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-13 makeStyles-root-17"
  >
    <header
      class="makeStyles-header-14"
    >
      <div
        class="makeStyles-toolbar-15"
      >
        <button
          class="makeStyles-button-18"
          tooltiptext="Toggle theme"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle theme"
            class="makeStyles-emoji-19 makeStyles-emoji-20"
            role="img"
          >
            ☀️
          </span>
        </button>
        <button
          class="makeStyles-button-18"
          tooltiptext="Toggle sound"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle sound"
            class="makeStyles-emoji-19 makeStyles-emoji-21"
            role="img"
          >
            🔔
          </span>
        </button>
      </div>
    </header>
    <main
      class="makeStyles-content-16"
    >
      <code
        class="makeStyles-time-22 makeStyles-time-23"
      >
        01:00
      </code>
      <button
        class="makeStyles-button-18"
      >
        <span
          aria-hidden="true"
          aria-label=""
          class="makeStyles-emoji-19 makeStyles-emoji-24"
          role="img"
        >
          work
        </span>
      </button>
    </main>
  </div>
</DocumentFragment>
`;

exports[`App renders correctly after emoji has been changed 2`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-13 makeStyles-root-17"
  >
    <header
      class="makeStyles-header-14"
    >
      <div
        class="makeStyles-toolbar-15"
      >
        <button
          class="makeStyles-button-18"
          tooltiptext="Toggle theme"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle theme"
            class="makeStyles-emoji-19 makeStyles-emoji-20"
            role="img"
          >
            ☀️
          </span>
        </button>
        <button
          class="makeStyles-button-18"
          tooltiptext="Toggle sound"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle sound"
            class="makeStyles-emoji-19 makeStyles-emoji-21"
            role="img"
          >
            🔔
          </span>
        </button>
      </div>
    </header>
    <main
      class="makeStyles-content-16"
    >
      <code
        class="makeStyles-time-22 makeStyles-time-23"
      >
        01:00
      </code>
      <button
        class="makeStyles-button-18"
      >
        <span
          aria-hidden="true"
          aria-label=""
          class="makeStyles-emoji-19 makeStyles-emoji-24"
          role="img"
        >
          👩‍💻
        </span>
      </button>
    </main>
  </div>
</DocumentFragment>
`;

exports[`App renders correctly after session type has been changed 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-25 makeStyles-root-29"
  >
    <header
      class="makeStyles-header-26"
    >
      <div
        class="makeStyles-toolbar-27"
      >
        <button
          class="makeStyles-button-30"
          tooltiptext="Toggle theme"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle theme"
            class="makeStyles-emoji-31 makeStyles-emoji-32"
            role="img"
          >
            ☀️
          </span>
        </button>
        <button
          class="makeStyles-button-30"
          tooltiptext="Toggle sound"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle sound"
            class="makeStyles-emoji-31 makeStyles-emoji-33"
            role="img"
          >
            🔔
          </span>
        </button>
      </div>
    </header>
    <main
      class="makeStyles-content-28"
    >
      <code
        class="makeStyles-time-34 makeStyles-time-35"
      >
        01:00
      </code>
      <button
        class="makeStyles-button-30"
      >
        <span
          aria-hidden="true"
          aria-label=""
          class="makeStyles-emoji-31 makeStyles-emoji-36"
          role="img"
        >
          work
        </span>
      </button>
    </main>
  </div>
</DocumentFragment>
`;

exports[`App renders correctly after session type has been changed 2`] = `
<DocumentFragment>
  <div
    class="makeStyles-root-25 makeStyles-root-29"
  >
    <header
      class="makeStyles-header-26"
    >
      <div
        class="makeStyles-toolbar-27"
      >
        <button
          class="makeStyles-button-30"
          tooltiptext="Toggle theme"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle theme"
            class="makeStyles-emoji-31 makeStyles-emoji-32"
            role="img"
          >
            ☀️
          </span>
        </button>
        <button
          class="makeStyles-button-30"
          tooltiptext="Toggle sound"
        >
          <span
            aria-hidden="false"
            aria-label="Toggle sound"
            class="makeStyles-emoji-31 makeStyles-emoji-33"
            role="img"
          >
            🔔
          </span>
        </button>
      </div>
    </header>
    <main
      class="makeStyles-content-28"
    >
      <code
        class="makeStyles-time-34 makeStyles-time-35"
      >
        01:00
      </code>
      <button
        class="makeStyles-button-30"
      >
        <span
          aria-hidden="true"
          aria-label=""
          class="makeStyles-emoji-31 makeStyles-emoji-36"
          role="img"
        >
          break
        </span>
      </button>
    </main>
  </div>
</DocumentFragment>
`;
